### PR TITLE
add receiver_adaptor utility to P2300R3

### PR DESCRIPTION
**Pull request by ericniebler**
refs #215

### CODE REVIEW

**Review by kirkshoop — COMMENTED** [review]
I noticed that retry doe not use the new receiver base class. Should it?

**Comment by ericniebler** [comment]
> I noticed that retry doe not use the new receiver base class. Should it?

For `retry`, the inner receiver needs to live in the operation state. So a `receiver_adaptor` that aggregates an inner receiver isn't very helpful. This could indicate that there is a more general form for a receiver adaptor than what is proposed.

**Comment by kirkshoop** [comment]
> This could indicate that there is a more general form for a receiver adaptor than what is proposed.

Yeah. I think many receivers must refer to a receiver stored in the operation state. Perhaps the general for is for the adapter to require the derived receiver to have a get_receiver() method

**Comment by ericniebler** [comment]
> Perhaps the general form is for the adapter to require the derived receiver to have a get_receiver() method

This is an interesting suggestion. I'll noodle on that, thanks.

File: examples/algorithms/retry.hpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+};
+
+// pass through all customizations except set_error, which retries the operation.
+template<class O, class R>
+struct _retry_receiver
+  : std::execution::receiver_adaptor<_retry_receiver<O, R>> {
+  O* o_;
+
+  R&& base() && noexcept { return (R&&) o_->r_; }
+  const R& base() const & noexcept { return o_->r_; }
```
Comment (kirkshoop) [inline examples/algorithms/retry.hpp:50]: Yeah. That looks good. I did confuse base with base class for a bit. If you want to try another name. The function of that receiver is destination or completion or result or output.

File: examples/retry.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+};
+
+int main() {
+  auto x = retry(fail_some{});
+  // prints:
+  //   fail!
+  //   fail!
+  //   success!
+  auto [a] = std::this_thread::sync_wait(std::move(x)).value();
```
Comment (kirkshoop) [inline examples/retry.cpp:61]: Should this validate or print`a`?

**Review by kirkshoop — COMMENTED** [review]
The complexity of supporting both storage in the adaptor and outside the adaptor seems a bit excessive when the derived receiver can store it, yes?

It looks like you might be updating the docs before switching this out of draft. So I'll just leave comments.

File: examples/algorithms/retry.hpp (hunk 1 above)
Comment (ericniebler) [reply examples/algorithms/retry.hpp:50]: All the iterator and view adaptors provide accessors named `base()` that return the inner thing.

**Comment by ericniebler** [comment]
> The complexity of supporting both storage in the adaptor and outside the adaptor seems a bit excessive when the derived receiver can store it, yes?

It's true, the derived type can always just store the receiver. But much of the complexity will be needed anyway for the other adaptors, which will almost certainly want the adaptor to store the inner object. All sender adaptors aggregate the inner sender, for instance.

Given that much of the machinery needs to exist anyway, I figured why not put it to good use here? It does simplify the `then` receiver somewhat.

**Comment by kirkshoop** [comment]
I don't consider it blocking. 

This design works, but is more complicated to specify and to discuss in LEWG. 

Derived senders and op states can also provide the storage so I fall on the other side of the complexity convenience trade off on this one.

**Comment by kirkshoop** [comment]
More as a laugh. It is also possible to take a lambda in the adaptor constructor which would return the base.  if that would seem lower impact than adding the base() functions to the derived.

File: examples/algorithms/retry.hpp (hunk 1 above)
Comment (kirkshoop) [reply examples/algorithms/retry.hpp:50]: Yeah. It looks weirder when base() is implemented in the derived class.

**Comment by ericniebler** [comment]
> More as a laugh. It is also possible to take a lambda in the adaptor constructor which would return the base. if that would seem lower impact than adding the base() functions to the derived.

Passing it to the constructor isn't enough. The rest of the adaptor needs to know the type of the base, too. Before settling on this design, I played with one where a lambda was specified as a template parameter. I didn't find it very satisfying, and it was a little cryptic.

**Comment by ericniebler** [comment]
> more complicated to specify and to discuss in LEWG

I have an idea for how this can be specified in a fairly straightforward manner.

**Review by kirkshoop — APPROVED** [review]

**Pull request merged**